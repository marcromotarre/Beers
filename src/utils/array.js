export const generateAutoIncrementalArrayFromMinToMax = (min, max) => {
  return new Array(max - min + 1).fill().map((number, index) => min + index);
};
