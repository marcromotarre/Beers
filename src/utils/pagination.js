export const getNumPages = ({ totalElements, elementsPerPage }) => {
  return Math.ceil(totalElements / elementsPerPage);
};

export const generatePageNumbers = ({
  actualPage,
  numPages,
  maxNumShowedPages,
}) => {
  const pagesPerSide = Math.floor(maxNumShowedPages / 2);
  const arrayPages = new Array(numPages)
    .fill()
    .map((number, index) => index + 1);

  if (numPages <= maxNumShowedPages) return { pages: arrayPages };
  if (actualPage - pagesPerSide <= 1) {
    return {
      last: numPages,
      pages: arrayPages.filter((page) => page < maxNumShowedPages - 1),
    };
  }
  if (actualPage + pagesPerSide >= numPages) {
    return {
      first: 1,
      pages: arrayPages.filter(
        (page) => page > numPages - (maxNumShowedPages - 2)
      ),
    };
  }

  return {
    first: 1,
    pages: arrayPages.filter(
      (page) =>
        page < actualPage + (pagesPerSide - 1) &&
        page > actualPage - (pagesPerSide - 1)
    ),
    last: numPages,
  };
};
