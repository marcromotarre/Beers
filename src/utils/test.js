import React from 'react';
import { render } from '@testing-library/react';

import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';

import { createStore } from 'redux';
import reducer from './../redux/reducers/index';
import { Provider } from 'react-redux';

export function renderRouteRedux(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <LocationProvider history={history}>{children}</LocationProvider>
      </Provider>
    );
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderRedux(
  ui,
  {
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderReduxWithRouter(
  ui,
  {
    route = '/',
    history = createHistory(createMemorySource(route)),
    initialState,
    store = createStore(reducer, initialState),
    ...renderOptions
  } = {}
) {
  return {
    ...render(
      <Provider store={store}>
        <LocationProvider history={history}>{ui}</LocationProvider>
      </Provider>
    ),
    history,
  };
}
