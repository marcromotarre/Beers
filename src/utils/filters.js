import { formatBrewed } from './text';
import { generateAutoIncrementalArrayFromMinToMax } from './array';

export const getBeersByBrewedFromTo = ({
  beers = [],
  brewedFrom,
  brewedTo,
}) => {
  return beers.filter((beer) => {
    const brewed = formatBrewed(beer.first_brewed);
    return (
      (brewed.year > brewedFrom.year && brewed.year < brewedTo.year) ||
      (brewed.year === brewedFrom.year &&
        brewed.month >= brewedFrom.month &&
        brewed.year < brewedTo.year) ||
      (brewed.year > brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month <= brewedTo.month) ||
      (brewed.year === brewedFrom.year &&
        brewed.year === brewedTo.year &&
        brewed.month >= brewedFrom.month &&
        brewed.month <= brewedTo.month)
    );
  });
};

export const getMinBrewedDate = (beers = []) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? -1 : 1;
    })[0];
};

export const getMaxBrewedDate = (beers) => {
  return beers
    .map((beer) => formatBrewed(beer.first_brewed))
    .sort(function (a, b) {
      return a.year <= b.year && a.month <= b.month ? 1 : -1;
    })[0];
};

export const getYearsFromMinToMaxBrewed = (beers) => {
  return generateAutoIncrementalArrayFromMinToMax(
    getMinBrewedDate(beers).year,
    getMaxBrewedDate(beers).year
  );
};

export const getMonths = () => {
  return generateAutoIncrementalArrayFromMinToMax(1, 12);
};
