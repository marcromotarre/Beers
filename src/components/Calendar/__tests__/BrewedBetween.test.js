import React from 'react';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';

import BrewedBetween from '../BrewedBetween';
import beers from '../../../fixtures/beers';
import { renderRedux } from '../../../utils/test';

/** this component should:
 *  render:
 *      - Calendar from & Calendar to (view and selected versions)
 *
 *  actions:
 *      - swipe from view and selected calendar version
 *      - dispatch setBrewedFrom when from selection calendar is closed
 *      - dispatch setBrewedTo when to selection calendar is closed
 *
 */

const initialState = {
  beersReducer: { pending: false, error: false, beers },
  filtersReducer: {
    brewedFrom: { month: 3, year: 2013 },
    brewedTo: { month: 5, year: 2017 },
  },
};

describe('BrewedBetween Component Snapshots', () => {
  it('BrewedBetween Snapshots', () => {
    expect(
      renderRedux(<BrewedBetween />, {
        initialState,
      }).baseElement
    ).toMatchSnapshot();
  });
});

describe('BrewedBetween', () => {
  it('BrewedBetween Component should render all information', () => {
    const { queryByTestId, getByTestId } = renderRedux(<BrewedBetween />, {
      initialState,
    });

    const brewedBetweenContainer = getByTestId('brewed-between-container');
    expect(brewedBetweenContainer.children.length).toBe(1);
    const brewedBetweenInfo = brewedBetweenContainer.children[0];
    expect(brewedBetweenInfo.children.length).toBe(3);

    expect(queryByTestId('calendar-month-year-view-2013-3')).toBeTruthy();
    expect(queryByTestId('calendar-month-year-view-2017-5')).toBeTruthy();
  });

  it('should change View Calendar to selection Calendar when click on it', () => {
    const { queryByTestId } = renderRedux(<BrewedBetween />, {
      initialState,
    });
    expect(queryByTestId('calendar-month-year-view-2013-3')).toBeTruthy();
    fireEvent.click(queryByTestId('calendar-month-year-view-2013-3'));
    expect(queryByTestId('calendar-month-year-view-2013-3')).toBeFalsy();
  });

  it('should change month when click on calendar, select diferent month and close it', () => {
    const { queryByTestId, getByTestId } = renderRedux(<BrewedBetween />, {
      initialState,
    });
    fireEvent.click(queryByTestId('calendar-month-year-view-2013-3'));

    const calendarMonthYearSelectionContainer = getByTestId(
      'calendar-month-year-selection-container'
    );

    const months = calendarMonthYearSelectionContainer.children[0];
    const january = months.children[0];
    fireEvent.click(january);

    expect(queryByTestId('calendar-month-year-view-2013-3')).toBeFalsy();
    expect(queryByTestId('calendar-month-year-view-2013-1')).toBeFalsy();
    fireEvent.click(
      queryByTestId('calendar-month-year-selection-check-button')
    );
    expect(queryByTestId('calendar-month-year-view-2013-3')).toBeFalsy();
    expect(queryByTestId('calendar-month-year-view-2013-1')).toBeTruthy();
  });
});
