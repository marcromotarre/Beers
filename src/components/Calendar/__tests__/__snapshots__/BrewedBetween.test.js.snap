// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrewedBetween Component Snapshots BrewedBetween Snapshots 1`] = `
<body>
  <div>
    <section
      class="sc-Axmtr jpHgIa"
      data-testid="brewed-between-container"
    >
      <section
        class="sc-AxmLO hshgKp"
      >
        <button
          data-testid="calendar-month-year-view-2013-3"
        >
          <section
            class="sc-AxjAm eKZXqA"
          >
            <div
              class="sc-AxiKw euRTWC"
              height="68.4931506849315px"
              width="100px"
            >
              <div
                class="sc-AxhUy hYaWOo"
                data-testid="calendar-month-year-view-container"
              >
                <label
                  class="sc-AxgMl gkdRyW"
                >
                  MARCH
                </label>
                <label
                  class="sc-AxheI beNytf"
                >
                  2013
                </label>
              </div>
            </div>
            <svg
              class="sc-AxirZ ktnnZK"
              width="100"
            />
          </section>
        </button>
        <div
          class="sc-fzozJi hXlPba"
        >
          <label>
            brewed
          </label>
          <svg
            width="28"
          />
          <label>
            between
          </label>
        </div>
        <button
          data-testid="calendar-month-year-view-2017-5"
        >
          <section
            class="sc-AxjAm eKZXqA"
          >
            <div
              class="sc-AxiKw euRTWC"
              height="68.4931506849315px"
              width="100px"
            >
              <div
                class="sc-AxhUy hYaWOo"
                data-testid="calendar-month-year-view-container"
              >
                <label
                  class="sc-AxgMl gkdRyW"
                >
                  MAY
                </label>
                <label
                  class="sc-AxheI beNytf"
                >
                  2017
                </label>
              </div>
            </div>
            <svg
              class="sc-AxirZ ktnnZK"
              width="100"
            />
          </section>
        </button>
      </section>
    </section>
  </div>
</body>
`;
